src/rest_vol_dataset.c

```diff
@@ -829,6 +834,23 @@ RV_dataset_write(size_t count, void *dset[], hid_t mem_type_id[], hid_t mem_spac
             FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_BADVALUE, FAIL, "memory datatype is invalid");
 
         write_body_len = (size_t)file_select_npoints * dtype_size;
+        if ((contiguous = RV_dataspace_selection_is_contiguous(mem_space_id[0])) < 0)
+            FUNC_GOTO_ERROR(H5E_DATASPACE, H5E_BADVALUE, FAIL,
+                            "Unable to determine if the dataspace selection is contiguous");
+        if (!contiguous) {
+            if (NULL == (write_body = (char *)RV_malloc(write_body_len)))
+                FUNC_GOTO_ERROR(H5E_DATASPACE, H5E_CANTALLOC, FAIL,
+                                "can't allocate space for the 'write_body' values");
+            if (H5Dgather(mem_space_id[0], buf[0], mem_type_id[0], write_body_len, write_body, NULL,
+                          write_body) < 0)
+                FUNC_GOTO_ERROR(H5E_DATASET, H5E_WRITEERROR, FAIL, "can't gather data to write buffer");
+            buf[0] = write_body;
+        }
+        else {
+            if ((offset = RV_convert_start_to_offset(mem_space_id[0])) < 0)
+                FUNC_GOTO_ERROR(H5E_DATASET, H5E_BADVALUE, FAIL, "Unable to determine memory offset value");
+            buf[0] = buf[0] + offset * dtype_size;
```

Comment:
What's the motivation for this? I think library functions should already take this into account, since they'll have access to the memory dataspace, and this could potentially lead to issues if the library expects the empty space at the start of the buffer.

Reply:
In the case of binary transfer, the variable `buf[0]` is used to set `uinfo.buffer` and it assumes that is the starting location in memory of the data to be transfer. The amount of data that needs to be transfer is specified by `uinfo.buffer_size` which is set to `write_body_len`. In the case of contiguous dataspace selection this starting point does not have to be at the input memory location of `buf[0]` but has some offset to it. Therefore, one needs to compute this offset and add it to `buf[0]`. This is done all to avoid unnecessary overhead as requested by @jhendersonHDF in PR #50. A way to visual this case is to consider a 1D example and make a selection which is contiguous but does not start at the first memory position.

If desired I can modify one of the hyperslab tests such that the offset is non-zero.

Reply:
Ah, I see. It would be best to include a test with a nonzero offset, since this is new functionality.